packages/backend/server/src/core/doc-renderer/controller.ts

```diff
@@ -0,0 +1,112 @@
+
+@Controller('/workspace/:workspaceId/:docId')
+export class DocRendererController {
+  constructor(
+    private readonly doc: DocManager,
+    private readonly permission: PermissionService,
+    private readonly url: URLHelper
+  ) {}
+
+  @Get()
+  async render(
+    @Res() res: Response,
+    @Param('workspaceId') workspaceId: string,
+    @Param('docId') docId: string
+  ) {
+    if (workspaceId === docId) {
+      throw new DocNotFound({ workspaceId, docId });
+    }
+
+    // if page is public, show all
+    // if page is private, but workspace public og is on, show og but not content
+    let og = false;
+    let content = false;
+    const isPagePublic = await this.permission.isPublicPage(workspaceId, docId);
+
+    if (isPagePublic) {
+      og = true;
+      content = true;
+    } else {
+      const allowPreview = await this.permission.allowUrlPreview(workspaceId);
+
+      if (allowPreview) {
+        og = true;
+      }
+    }
+
+    if (og) {
+      const docContent = await this.doc.getPageContent(workspaceId, docId);
+
+      res
+        .setHeader('Content-Type', 'text/html')
+        .send(
+          this._render(
+            workspaceId,
+            docContent ?? { title: '', summary: '', content: [] },
+            content
+          )
+        );
+    } else {
+      throw new DocNotFound({ workspaceId, docId });
+    }
+  }
+
+  _render(workspaceId: string, doc: PageDocContent, content: boolean): string {
+    return `
+      <!DOCTYPE html>
+      <html>
+        <head>
+          <title>${doc.title} | AFFiNE</title>
+          <meta name="theme-color" content="#fafafa" />
+          <link rel="manifest" href="/manifest.json" />
+          <link rel="apple-touch-icon" sizes="180x180" href="/apple-touch-icon.png" />
+          <link rel="icon" sizes="192x192" href="/favicon-192.png" />
+          <meta
+            name="twitter:title"
+            content="AFFiNE: There can be more than Notion and Miro."
+          />
+          <meta name="twitter:description" content="${doc.title}" />
+          <meta name="twitter:site" content="@AffineOfficial" />
+          <meta name="twitter:image" content="https://affine.pro/og.jpeg" />
+          <meta property="og:title" content="${doc.title}" />
+          <meta property="og:description" content="${doc.summary}" />
+          <meta property="og:image" content="https://affine.pro/og.jpeg" />
+        </head>
+        <body>
+          <header><h1>${doc.title}</h1></header>
+          <main>
+            ${content ? doc.content.map(block => this._renderContent(workspaceId, block)).join('') : `<p>${doc.summary}</p>`}
+          </main>
+        </body>
+      </html>
+    `;
+  }
+
+  _renderContent(workspaceId: string, content: PageDocContentType): string {
+    switch (content.type) {
+      case 'affine:paragraph':
+        return content.variant === 'text'
+          ? `<p>${content.text}</p>`
+          : `<${content.variant}>${content.text}</${content.variant}>`;
```

Comment:
based on the same reasons as before, can we add the `rel=nofollow` attribute to all external links?

Reply:
no link rendered currently